README.md
## About [![build status](https://secure.travis-ci.org/SaltwaterC/ftp-get.png?branch=master)](http://travis-ci.org/SaltwaterC/ftp-get) ![still maintained](http://stillmaintained.com/SaltwaterC/ftp-get.png)

Simple FTP client for node.js implemented in pure JavaScript. Useful for downloading files from a remote location, therefore it implements just a small subset of the FTP protocol. Includes a method, modeled after HTTP's HEAD in order to check the existence of a remote resource without downloading its contents. The ftp.head() method uses the FTP SIZE command which a RFC 3659 extension to the protocol, since 2007. However, in practice most FTP servers implement the SIZE extension before the RFC publishing date.

## Installation

Either manually clone this repository into your node_modules directory, or the recommended method:

> npm install ftp-get

## Usage mode

 * The [ftp.get method](https://github.com/SaltwaterC/ftp-get/wiki/ftp.get-method)
 * The [ftp.head method](https://github.com/SaltwaterC/ftp-get/wiki/ftp.head-method)

## Bug reporting

Unlike 0.2 and before, the 0.3 version of ftp-get went through extensive testing with real world FTP servers. However, some edge cases may still fail. Therefore, in order to have a proper bug repot I'm kindly asking you to provide the troublesome FTP URL (if possible).

I'd appreciate the output of this script as well:

```bash
NODE_ENV=development node client.js ftp://example.com/ftp/url.foo
```

```javascript
// client.js
var ftp = require('ftp-get')

var url = process.argv[2]

ftp.get(url, 'dl.bin', function (err, res) {
	console.log(err, res)
})
```


CHANGELOG.md
## v0.3.1
 * Changes the internal buffering concatenation to a much faster implementation: Buffer.concat by using the buffer-concat module for node.js v0.6.

## v0.3.0
 * Client rewritten (almost) from scratch.
 * Better debug mode.
 * node.js v0.6+ support only.
 * jslint compliant.
 * Support for Buffer buffering. The String buffers are deprecated.
 * New error codes. Removed a couple of old error codes.
 * Configurable timeouts.
 * Battle hardened.

## v0.2.8
 * Fixes the ftp.head() not calling the callback when the underlying socket closes prematurely. Uses the same patch for ftp.get() in order to fail faster than waiting for the data socket to timeout.
 * The close / error events for both ftp.get() and ftp.head() both call the callback with the same set of properties.

## v0.2.7
 * Fixes the ftp.head() for when socket errors are emitted.

## v0.2.6
 * Adds support for ftp.head() method modeled after the idea of HTTP's HEAD.

## v0.2.5
 * Fixes a possible race condition where a single data event for the disk transfer case might not triggger the end callback.

## v0.2.4
 * Fixes a race condition where a single data event from the data socket could be missed.

## v0.2.3
 * Removed the last QUIT command before closing the command channel.

## v0.2.2
 * Makes sure that the success isn't trigered for the saved to the disk transfer if there's an error.

## v0.2.1
 * Adds a [fsync(2)](http://linux.die.net/man/2/fsync) wrapper for the transfers that are saved to disk, improving the library stability.

## v0.2
 * Bugfix: Ignores repeated FTP errors. Calls the callback only once. Implemented a flag that keeps the error callback away, therefore it removes any race condition.
 * Enhancement: Adds error codes for each failure as well as debug information into the error argument.
 * Enhancement: Made the data listener after connection to wait for all the data events that do not end with CRLF. This is a cleaner approact that removes the timeouts for the servers that don't send their welcome junk in a single package.
 * Changes the returned value of the err argument in the success case from undefined to null in order to follow exactly the node.js convention. This may break some code if the evaluation is made against 'undefined'.

## v0.1.5
 * Rewrite most of the library internals.

## v0.1.4
 * Introduced some crappy timeouts just to please the Microsoft FTP Service when doing socket.write().

## v0.1.3
 * Trims the whitespace at the beginning and at the end of the input URL.
 * Trims the whitespace at the beginning and at the end of the input file path.

## v0.1.2
 * Small optimization: if the second argument is a callback, the path.resolve() call is avoided.
 * Declares the socket variable into the get method.

## v0.1.1
 * Added support for buffering the FTP file instead of saving it to the disk. Useful for things that need to be processed without the need of saving the file to the disk.

## v0.1
 * Initial version, featuring support for fetching FTP files from a remote location into a file onto your local machine.


tests/file.js
'use strict';

var ftp = require('../');

var assert = require('assert');
var fs = require('fs');

var common = require('./includes/common.js');

var file = './foo.txt';
var callbacks ={
	get: 0
};

ftp.get('ftp://127.0.0.1/foo.txt', file, function (err, res) {
	callbacks.get++;
	
	assert.ifError(err);
	fs.stat(file, function (err) {
		assert.ifError(err);
		fs.readFile(file, function (err, data) {
			assert.ifError(err);
			assert.strictEqual(data.toString(), 'bar\n');
			fs.unlink(file, function (err) {
				assert.ifError(err);
			});
		});
	});
});

common.teardown(callbacks);


tests/head.js
'use strict';

var ftp = require('../');

var assert = require('assert');

var common = require('./includes/common.js');

var callbacks = {
	head: 0
};

ftp.head('ftp://127.0.0.1/foo.txt', function (err, size) {
	callbacks.head++;
	assert.ifError(err);
	assert.strictEqual(size, 4);
});

common.teardown(callbacks);


tests/close.js
'use strict';

var ftp = require('../');

var net = require('net');
var util = require('util');
var assert = require('assert');

var common = require('./includes/common.js');

var callbacks = {
	get: 0,
	head: 0
};
var expect = 2;

var url = 'ftp://127.0.0.1:2121/foo.txt';

util.log('running on node.js ' + process.version);

var server = net.createServer(function (conn) {
	conn.setEncoding('utf8');
	
	conn.on('data', function (data) {
		util.log('being rude and closing the command socket');
		conn.end();
	});
	
	util.log('sending the server hello message');
	conn.write('220 Hello random people I do not care about' + ftp.CRLF);
}).listen(2121, function () {
	util.log('server listening on 2121');
	
	var assertions = function (err, res) {
		assert.ok(err instanceof Error);
		assert.strictEqual(err.code, 11);
		
		if (expect === 0) {
			util.log('closing the rude server');
			server.close();
		}
	};
	
	ftp.get({
		url: url,
		bufferType: 'buffer'
	}, function (err, res) {
		util.log('ftp.get');
		callbacks.get++;
		expect--;
		assertions(err, res);
	});
	
	ftp.head(url, function (err, res) {
		util.log('ftp.head');
		callbacks.head++;
		expect--;
		assertions(err, res);
	});
});

common.teardown(callbacks);


lib/ftp-get.js
'use strict';

/* core modules */
var fs = require('fs');
var u = require('url');
var p = require('path');
var net = require('net');
var util = require('util');
var EventEmitter = require('events').EventEmitter;

/* 3rd party module */
require('buffer-concat'); // patches in Buffer.concat for node.js v0.6.x

var CRLF = '\r\n';
exports.CRLF = CRLF;

/**
 * util.debug wrapper. Outputs only when NODE_ENV=development
 *
 * @param {String} message
 */
var debug = function (message) {
	if (process.env.NODE_ENV === 'development') {
		util.debug('ftp-get - ' + String(message));
	}
};

/**
 * Returns the absolute integer value of the input. Avoids the NaN crap.
 * 
 * @param value
 * @returns {Number}
 */
var absInt = function (value) {
	/*jslint bitwise:true*/
	var sureInt = parseInt(value, 10) | 0;
	/*jslint bitwise:false*/
	return Math.abs(sureInt);
};

/**
 * Authentication info parser
 *
 * @param {String} auth
 * @returns {Object} auth
 */
var parseAuth = function (auth) {
	if (auth) {
		auth = auth.split(':');
		if (auth[0] && auth[1]) {
			return {
				user: auth[0],
				pass: auth[1]
			};
		}
	}
	
	return {};
};

/**
 * URL parser
 *
 * @param {String} url
 * @returns {Object} url
 */
var processUrl = function (url) {
	url = String(url);
	url = url.trim();
	url = u.parse(url);
	url.port = url.port || 21;
	
	var auth = parseAuth(url.auth);
	url.user = auth.user || 'anonymous';
	url.pass = auth.pass || 'anonymous@';
	
	// in order to save a CWD round trip
	// stripping the forward slash for files
	// at the apex of the FTP list
	// wu-ftpd + anonymous fails on these
	var path = url.pathname.split('/');
	if (path.length === 2) {
		url.pathftp = url.pathname.substring(1);
	} else {
		url.pathftp = url.pathname;
	}
	
	// ftp breaks on %20 and other stuff like this
	url.pathftp = decodeURIComponent(url.pathftp);
	
	return url;
};

/**
 * Parses the FTP server messages
 *
 * @param {String} data
 * @param {Object} messages
 */
var parseMessages = function (data) {
	var idx;
	var messages = [];
	
	data = String(data).split(CRLF);
	
	for (idx in data) {
		if (data.hasOwnProperty(idx)) {
			var code = data[idx].substring(0, 3);
			var message = data[idx].substring(3);
			
			messages.push({
				code: code,
				message: message
			});
			
			if (code !== '') {
				debug('socket.receive ' + code + ': ' + message);
			}
		}
	}
	
	return messages;
};

var endOfMessages = function (messages) {
	var idx;
	
	for (idx in messages) {
		if (messages.hasOwnProperty(idx)) {
			// some servers don't send the complete reply in one packet
			// node splits it in different data events
			// have fun debugging that ...
			if (messages[idx].code === '') {
				return true;
			}
		}
	}
	
	return false;
};

/**
 * The FTP client
 *
 * @param {Object} opt
 */
function FTP(opt) {
	EventEmitter.call(this);
	
	if (typeof opt === 'string') {
		opt = {url: opt};
	}
	
	if (typeof opt !== 'object') {
		throw new Error('You must pass an object for the client options.');
	}
	
	if (opt.url === undefined) {
		throw new Error('You must pass an url option to the ftp-get library.');
	}
	
	opt.url = processUrl(opt.url);
	
	opt.timeout = absInt(opt.timeout);
	if (opt.timeout === 0) {
		opt.timeout = 60;
	}
	opt.timeout = opt.timeout * 1000;
	
	opt.bufferType = opt.bufferType || 'string';
	if (opt.bufferType !== 'buffer' && opt.bufferType !== 'string') {
		throw new Error('Expecting buffer or string as buffer type.');
	}
	
	this.opt = opt;
	
	this.error = false;
	this.ended = false;
}

util.inherits(FTP, EventEmitter);

/**
 * Watches the data events for FTP errors
 *
 * @param {Array} messages
 * @returns {Object} error
 */
FTP.prototype.errorWatch = function (messages) {
	var self = this;
	var idx;
	
	for (idx in messages) {
		if (messages.hasOwnProperty(idx)) {
			if (messages[idx].code.charAt(0) === '4' || messages[idx].code.charAt(0) === '5') {
				if ( ! self.error && ! self.ended) {
					self.cleanup();
					
					var err = new Error(util.format('FTP error %d', messages[idx].code));
					err.code = Number(messages[idx].code);
					err.url = self.opt.url;
					self.emit('error', err);
					
					return err;
				}
			}
		}
	}
	
	return null;
};

/**
 * Makes sure the FTP client receives the proper connection message
 */
FTP.prototype.connect = function () {
	var self = this;
	
	self.cmdSock = net.createConnection(self.opt.url.port, self.opt.url.hostname);
	self.cmdSock.setEncoding('utf8');
	
	self.cmdSock.setTimeout(self.opt.timeout, function () {
		if ( ! self.error && ! self.ended) {
			self.cleanup();
			
			var err = new Error('command socket timeout');
			err.code = 7;
			err.url = self.opt.url;
			
			self.emit('error', err);
		}
	});
	
	self.cmdSock.on('error', function (err) {
		if ( ! self.error && ! self.ended) {
			self.cleanup();
			err.code = 5;
			err.url = self.opt.url;
			self.emit('error', err);
		}
	});
	
	self.cmdSock.on('close', function () {
		if ( ! self.ended) {
			self.cleanup();
			
			var err = new Error('command socket prematurely closed');
			err.code = 11;
			err.url = self.opt.url;
			
			self.emit('error', err);
		}
	});
	
	self.cmdSock.on('connect', function () {
		debug('connected to the FTP server on ' + self.opt.url.hostname + ':' + self.opt.url.port + '.');
		
		var ok = false;
		var end = false;
		
		var connect = function (data) {
			var messages = parseMessages(data);
			if ( ! (self.errorWatch(messages) instanceof Error)) {
				if (messages[0].code === '220') {
					ok = true;
				}
				
				end = endOfMessages(messages);
				
				if (end === true) {
					self.cmdSock.removeAllListeners('data');
					
					if (ok === false) {
						if ( ! self.error && ! self.ended) {
							self.cleanup();
							
							var err = new Error(util.format('ftp error %s', messages[0].code));
							err.code = Number(messages[0].code);
							err.url = self.opt.url;
							
							self.emit('error', err);
						}
					} else {
						self.auth();
					}
				}
			}
		};
		
		self.cmdSock.on('data', connect);
	});
};

/**
 * FTP auth wrapper
 */
FTP.prototype.auth = function () {
	var self = this;
	
	var loggedIn = false;
	var end = false;
	var pass = false;
	
	var login = function (data) {
		var idx;
		var messages = parseMessages(data);
		
		if ( ! (self.errorWatch(messages) instanceof Error)) {
			for (idx in messages) {
				if (messages.hasOwnProperty(idx)) {
					if (messages[idx].code === '230') {
						// user logged in
						end = true;
						pass = true;
					}
					
					if (messages[idx].code === '331') {
						// send the password
						end = true;
					}
				}
			}
			
			if (end === true) {
				self.cmdSock.removeAllListeners('data');
				
				if (pass === false) {
					self.cmdWrite('PASS ' + self.opt.url.pass, login);
				} else {
					self.type();
				}
			}
		}
	};
	
	self.cmdWrite('USER ' + self.opt.url.user, login);
};

/**
 * TYPE command wrapper
 */
FTP.prototype.type = function () {
	var self = this;
	var end = false;
	
	var type = function (data) {
		var idx;
		var messages = parseMessages(data);
		
		if ( ! (self.errorWatch(messages) instanceof Error)) {
			for (idx in messages) {
				if (messages.hasOwnProperty(idx)) {
					if (messages[idx].code === '200') {
						end = true;
					}
				}
			}
			
			if (end === true) {
				self.cmdSock.removeAllListeners('data');
				self.size();
			}
		}
	};
	
	self.cmdWrite('TYPE I' , type);
};

/**
 * SIZE command wrapper
 */
FTP.prototype.size = function () {
	var self = this;
	
	var code, message;
	var end = false;
	
	var size = function (data) {
		var messages = parseMessages(data);
		
		code = messages[0].code;
		message = messages[0].message;
		
		end = endOfMessages(messages);
		
		if (end === true) {
			self.cmdSock.removeAllListeners('data');
			switch (code) {
				case '213': // ok
					self.emit('size', message.trim());
				break;
				
				case '500': // SIZE extension of RFC 3659 not implemented
					self.emit('size', -1);
				break;
				
				default: // FTP error
					if ( ! self.error && ! self.ended) {
						self.cleanup();
						var err = new Error(util.format('ftp error %s', code));
						err.code = Number(code);
						err.url = self.opt.url;
						
						self.emit('error', err);
					}
				break;
			}
		}
	};
	
	self.cmdWrite('SIZE ' + self.opt.url.pathftp, size);
};

/**
 * Downloads a file from remote location
 */
FTP.prototype.download = function () {
	var self = this;
	var end = false;
	
	var pasv = function (data) {
		var messages = parseMessages(data);
		if ( ! (self.errorWatch(messages) instanceof Error)) {
			if (messages[0].code === '227') {
				var address = messages[0].message.match(/\(([\w\W]*)\)/);
				address = address[1];
				address = address.split(',');
				
				var host = address[0] + '.' + address[1] + '.' + address[2] + '.' + address[3];
				var port = (address[4] * 256) + absInt(address[5]);
				
				self.dataSock = net.createConnection(port, host);
				self.dataSock.setTimeout(self.opt.timeout, function () {
					if ( ! self.error && ! self.ended) {
						self.cleanup();
						var err = new Error('data socket timeout');
						err.code = 4;
						err.url = self.opt.url;
						self.emit('error', err);
					}
				});
				
				self.dataSock.on('error', function (err) {
					if ( ! self.error && ! self.ended) {
						self.cleanup();
						err.code = 3;
						err.url = self.opt.url;
						self.emit('error', err);
					}
				});
				
				self.dataSock.once('connect', function () {
					self.dataSock.pause();
					self.retr();
				});
			} else { // did not get the data socket information
				if ( ! self.error && ! self.ended) {
					self.cleanup();
					var error = new Error('could not get the data socket information');
					error.code = 9;
					error.url = self.opt.url;
					self.emit('error', error);
				}
			}
		}
		
		end = endOfMessages(messages);
		if (end === true) {
			self.cmdSock.removeAllListeners('data');
		}
	};
	
	self.cmdWrite('PASV', pasv);
};

/**
 * RETR command wrapper
 */
FTP.prototype.retr = function () {
	var self = this;
	var end = false;
	
	var retr = function (data) {
		var idx;
		var messages = parseMessages(data);
		
		if ( ! (self.errorWatch(messages) instanceof Error)) {
			for (idx in messages) {
				if (messages.hasOwnProperty(idx)) {
					if (messages[idx].code === '150' || messages[idx].code === '125') {
						end = true;
					}
				}
			}
			
			if (end === true) {
				self.emit('stream', self.dataSock);
				self.cmdSock.setTimeout(0);
				self.cmdSock.removeAllListeners('data');
			}
		}
	};
	
	self.cmdWrite('RETR ' + self.opt.url.pathftp, retr);
};

/**
 * Writes commands to the command socket
 *
 * @param {String} cmd
 * @param {Function} cb
 */
FTP.prototype.cmdWrite = function (cmd, cb) {
	var self = this;
	
	if (self.cmdSock.writable) {
		debug('socket.write: ' + cmd);
		self.cmdSock.write(cmd + CRLF, 'utf8', function () {
			self.cmdSock.on('data', function (data) {
				cb(data);
			});
		});
	}
};

/**
 * Cleaning up the client junk on error
 */
FTP.prototype.cleanup = function () {
	this.error = true;
	this.ended = true;
	
	if (this.cmdSock) {
		this.cmdSock.destroy();
	}
	
	if (this.dataSock) {
		this.dataSock.destroy();
	}
	
	if (this.writeStream) {
		this.writeStream.destroy();
	}
	
	if (this.file) {
		fs.unlink(this.file);
	}
};

/**
 * Clean close of all the client resources
 */
FTP.prototype.close = function () {
	this.ended = true;
	this.cmdSock.end();
};

/**
 * ftp.head method similar to HTTP's HEAD
 * Based on the SIZE command, therefore it fails for
 * older servers that don't implement RFC 
 *
 * @param {Object} opt
 * @param {Function} cb
 */
exports.head = function (opt, cb) {
	var client = new FTP(opt);
	
	client.on('error', function (err) {
		cb(err);
	});
	
	client.on('size', function (size) {
		if (size === -1) {
			var err = new Error('SIZE command not implemented');
			err.code = 8;
			err.url = client.opt.url;
			cb(err);
		} else {
			cb(null, Number(size));
		}
		
		client.close();
	});
	
	client.connect();
};

/**
 * ftp.get method for downloading a remote file
 *
 * @param {Object} opt
 * @param {String} file
 * @param {Function} cb
 */
exports.get = function (opt, file, cb) {
	if (typeof file === 'function') {
		cb = file;
		file = null;
	} else {
		file = String(file);
		file = file.trim();
		file = p.resolve(file);
	}
	
	var client = new FTP(opt);
	var fileSize = -1;
	var error = false;
	
	client.on('error', function (err) {
		if ( ! error) {
			error = true;
			cb(err);
		}
	});
	
	client.on('size', function (size) {
		debug('got the size event');
		fileSize = Number(size);
		client.download();
	});
	
	client.on('stream', function (stream) {
		debug('got the stream event');
		stream.resume();
		
		if (file === null) { // buffer the data
			var buf = [], size = 0;
			
			stream.on('error', function (err) {
				if ( ! error) {
					error = true;
					client.cleanup();
					
					err.code = 3; // data socket error
					err.url = client.opt.url;
					
					cb(err);
				}
			});
			
			stream.on('data', function (data) {
				size += data.length;
				
				// SlowBuffer is limited to 0x3FFFFFFF bytes aka 1073741823
				if (size > 1073741823) {
					error = true;
					client.cleanup();
					
					var err = new Error('Buffer too large. node.js limits the SlowBuffer implementation to 1073741823 bytes.');
					err.code = 13;
					err.url = client.opt.url;
					
					cb(err);
				}
				
				if ( ! error) {
					buf.push(data);
				}
			});
			
			stream.on('close', function () {
				if ( ! error) {
					var buffer = Buffer.concat(buf, size);
					
					if (fileSize !== -1 && fileSize !== buffer.length) {
						error = true;
						client.cleanup();
						
						var err = new Error(util.format('Wrong file size, expected %d, got %d', fileSize, buffer.length));
						err.code = 10;
						err.url = client.opt.url;
						
						cb(err);
					} else {
						if (client.opt.bufferType === 'string') {
							buffer = buffer.toString();
							console.error('The string bufferType is deprecated. Use the buffer bufferType');
						}
						
						cb(null, buffer);
						client.close();
					}
				}
			});
		} else { // save the data to disk
			var writeStream = fs.createWriteStream(file);
			var transfer = new EventEmitter();
			
			writeStream.on('error', function (err) {
				if ( ! error) {
					error = true;
					client.cleanup();
					
					error = true;
					err.code = 1;
					err.url = client.opt.url;
					
					cb(err);
				}
			});
			
			transfer.on('end', function () {
				transfer.removeAllListeners('end');
				writeStream.on('open', function (fd) {
					fs.fsync(fd, function () {
						writeStream.end();
					});
				});
			});
			
			writeStream.on('open', function (fd) {
				transfer.on('end', function () {
					transfer.removeAllListeners('end');
					fs.fsync(fd, function () {
						writeStream.end();
					});
				});
			});
			
			stream.on('close', function () {
				if ( ! error) {
					client.close();
					transfer.emit('end');
				}
			});
			
			writeStream.on('close', function () {
				if ( ! error) {
					fs.stat(file, function (err, stats) {
						var err2;
						if ( ! error) {
							if (err) {
								error = true;
								client.cleanup();
								
								err2 = new Error('the file does not exist at the end of the transfer');
								err2.code = 12;
								err2.url = client.opt.url;
								cb(err2);
							} else {
								if (fileSize !== -1 && fileSize !== stats.size) {
									error = true;
									client.cleanup();
									
									err2 = new Error(util.format('wrong file size, expected %d, got %d', fileSize, stats.size));
									err2.code = 10;
									err2.url = client.opt.url;
									
									cb(err2);
								} else {
									cb(null, file);
								}
							}
						}
					});
				}
			});
			
			stream.on('data', function (data) {
				try {
					writeStream.write(data);
				} catch (e) {} // handled by the error listener, sad that async methods throw
			});
		}
	});
	
	client.connect();
};
